lineoperations/src/linefunctions.c

```diff
@@ -0,0 +1,401 @@
+		sci_set_text(doc->editor->sci, newfile);	// set new document
+	}
+
+
+	// each line is freed in above for-loop
+	if(lines)    g_free(lines);
+	if(newfile)  g_free(newfile);
+	if(toremove) g_free(toremove);
+}
+
+
+
+void rmemtyln(GeanyDocument *doc) {
+	gint  totalnumchars;    // number of characters in the document
+	gint  totalnumlines;    // number of lines in the document
+	gchar *line;            // temporary line
+	gint  linelen;          // length of *line
+	gchar *newfile;         // *final* string to replace current document
+	gint  nfposn;           // position to the last character in newfile
+	gint  i;                // temporary iterator number
+	gint j;                 // iterator
+
+
+	totalnumchars = sci_get_length(doc->editor->sci);
+	totalnumlines = sci_get_line_count(doc->editor->sci);
+	/*
+	 * Allocate space for the new document (same amount as open document.
+	 * If the document contains lines with only whitespace,
+	 * not all of this space will be used.)
+	*/
+	newfile   = g_malloc(sizeof(gchar) * (totalnumchars+1));
+	nfposn    = 0;
+
+	if(newfile)    // verify memory allocation
+	{
+		for(i = 0; i < totalnumlines; i++)    // loop through opened doc char by char
+		{
+			linelen = sci_get_line_length(doc->editor->sci, i);
+
+			if(linelen == 2)
+			{
+				line = sci_get_line(doc->editor->sci, i);
+
+				if(line[0] != '\r')
+					// copy current line into *newfile
+					for(j = 0; line[j] != '\0'; j++)
+						newfile[nfposn++] = line[j];
+			}
+			else if(linelen != 1)
+			{
+				line = sci_get_line(doc->editor->sci, i);
+
+				// copy current line into *newfile
+				for(j = 0; line[j] != '\0'; j++)
+					newfile[nfposn++] = line[j];
+			}
+		}
+		newfile[nfposn] = '\0';
+		sci_set_text(doc->editor->sci, newfile);	// set new document
+
+
+		g_free(newfile);
+	}
+}
+
+
+
+// Remove Whitespace Lines
+void rmwhspln(GeanyDocument *doc) {
+	gint  totalnumchars;    // number of characters in the document
+	gint  totalnumlines;    // number of lines in the document
+	gchar *line;            // temporary line
+	gint  linelen;          // length of *line
+	gchar *newfile;         // *final* string to replace current document
+	gint  nfposn;           // position to the last character in newfile
+	gint  i;                // temporary iterator number
+	gint j;                 // iterator
+
+
+	totalnumchars = sci_get_length(doc->editor->sci);
+	totalnumlines = sci_get_line_count(doc->editor->sci);
+	/*
+	 * Allocate space for the new document (same amount as open document.
+	 * If the document contains lines with only whitespace,
+	 * not all of this space will be used.)
+	*/
+	newfile   = g_malloc(sizeof(gchar) * (totalnumchars+1));
+	nfposn    = 0;
+
+	if(newfile)    // verify memory allocation
+	{
+		for(i = 0; i < totalnumlines; i++)    // loop through opened doc
+		{
+			linelen = sci_get_line_length(doc->editor->sci, i);
+
+			if(linelen == 2)
+			{
+				line = sci_get_line(doc->editor->sci, i);
+
+				if(line[0] != '\r')
```

Comment:
The function `iswhitespaceline()` have been removed. It is no longer needed because of your suggestions with `sci_get_line_indentation` and direct scintilla manipulation.